src/Confluent.Kafka/Consumer.cs

```diff
@@ -220,19 +270,31 @@ private void LogCallback(IntPtr rk, SyslogLevel level, string fac, string buf)
                     }
 
                     lock (assignCallCountLockObj) { assignCallCount = 0; }
-                    var assignTo = partitionsRevokedHandler(assignmentWithPositions);
+                    var assignTo = kafkaHandle.AssignmentLost
+                        ? (partitionsLostHandler != null
+                            ? partitionsLostHandler(assignmentWithPositions)
+                            : partitionsRevokedHandler(assignmentWithPositions))
+                        : partitionsRevokedHandler(assignmentWithPositions);
                     lock (assignCallCountLockObj)
                     {
                         if (assignCallCount > 0)
                         {
-                            throw new InvalidOperationException("Assign/Unassign must not be called in the partitions revoked handler.");
+                            throw new InvalidOperationException("Assign/Unassign must not be called in the partitions revoked handler");
                         }
                     }
 
-                    // This distinction is important because calling Assign whilst the consumer is being
-                    // closed (which will generally trigger this callback) is disallowed.
-                    if (assignTo.Count() > 0) { Assign(assignTo); }
-                    else { Unassign(); }
+                    if (kafkaHandle.RebalanceProtocol == "COOPERATIVE")
+                    {
+                        // assignTo is always empty, not used in the COOPERATIVE case.
+                        IncrementalUnassign(partitions);
+                    }
+                    else
+                    {
+                        // This distinction is important because calling Assign whilst the consumer is being
```

Comment:
It is actually okay now, we just call unassign() internally.

Reply:
changing this to just call `Unassign`.